harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -207,6 +383,103 @@ def _set_data_loaders(self) -> None:
+
+        if issubclass(type(self.max_length), Batch):
+            self._train_step()
+
+    def _step_epoch(self):
+        self.epochs_trained += 1
+        if issubclass(type(self.max_length), Epoch):
+            self._train_step()
+
+    def _train_step(self):
+        self.steps_completed += 1
+        if self.steps_completed % self.validation_period == 0:
+            self.report_training_progress(self.searcher_op)
+            val_metrics = self._validate()
+            self.report_validation_metrics(self.steps_completed, val_metrics)
+            self.upload_tb_files()
+            self.stop_requested()
+        if self.steps_completed % self.checkpoint_period == 0:
+            self.report_training_progress(self.searcher_op)
+            self.checkpoint(already_exiting=False)
+            self.upload_tb_files()
+            self.stop_requested()
+
+    def stop_requested(self):
+        if self.core_context.distributed.rank == 0:
+            if self.core_context.preempt.should_preempt():
+                raise ShouldExit()
+        if self.context.get_stop_requested():
+            raise ShouldExit()
+
+    def _steps_remaining(self):
+        return self.max_length.value - self.steps_completed
+
+    def _train(self, train_unit, min_checkpoint_period, min_validation_period):
+        self.max_length = train_unit
+        self.validation_period = min_validation_period
+        self.checkpoint_period = min_checkpoint_period
+
+        while self._steps_remaining() > 0:
+            for batch_idx, batch in enumerate(self.training_loader):
+                training_metrics = self._train_batch(self.batches_trained, batch,
+                                                     self.epochs_trained, batch_idx)
+                self.training_metrics.append(training_metrics)
+                self._step_batch()
+                if not self._steps_remaining() > 0:
+                    return self.training_metrics
+            self._step_epoch()
+
+        return self.training_metrics
+
+    def report_training_progress(self, op):
+        searcher_unit = self.core_context.searcher.get_configured_units()
+        if self.core_context.distributed.rank == 0:
+            if issubclass(type(self.max_length), Batch):
+                if searcher_unit == core.Unit.BATCHES:
+                    op.report_progress(self.steps_completed)
+                elif searcher_unit == core.Unit.RECORDS:
+                    op.report_progress(self.context.get_global_batch_size() * self.steps_completed)
+                elif searcher_unit == core.Unit.EPOCHS:
+                    op.report_progress(self.steps_completed / self.as_batches(epochs=1))
+                else:
+                    raise ValueError(f"unrecognized searcher op unit: {searcher_unit}")
+            else:
+                raise ValueError(f"unsupported training unit {type(self.max_length)}")
+        self.report_training_metrics(self.steps_completed)
+
+    def as_batches(
```

Comment:
ditto; I think `as_batches()` should not exist after wlsq is gone... otherwise wlsq isn't really gone, it's just been copied/pasted/merged into PyTorchTrial.

Reply:
agree, but this has to be done with supporting true epoch based training. this refactor is largely to support a trainer API with the same internal logic as existing PyTorchTrial.